Refactor Storage Create/Edit/Empty/Delete Modals to use Shadcn components

Saeris opened the pull request:
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.

YES

## What kind of change does this PR introduce?

Following the same implementation patterns as in #37258, #37308, and #37386, this PR refactors `CreateBucketModal`, `EditBucketModal`, `EmptyBucketModal`, and `DeleteBucketModal` to use the shadcn based form controls and dialog patterns. It also cleans up `StorageMenu` and `BucketRow` to reduce the overall complexity involved in the usage of these modals. Finally, it adds some basic tests for each of these modals.

Some before/after screenshots:

`CreateBucketModal`
<img width="2552" height="1344" alt="image" src="https://github.com/user-attachments/assets/cc0ac70d-71ca-4e77-b2e1-381840c179d9" />

`EditBucketModal`
<img width="2554" height="1340" alt="image" src="https://github.com/user-attachments/assets/9f1db2ba-fc13-48fa-90c3-6f490e6e39ca" />

`EmptyBucketModal`
<img width="2555" height="1344" alt="image" src="https://github.com/user-attachments/assets/bc50e418-006e-4220-a3bd-91cdd8777a42" />

`DeleteBucketModal`
<img width="2554" height="1341" alt="image" src="https://github.com/user-attachments/assets/dc62b77f-cb42-472d-ae51-740ce1c28b97" />

Saeris committed:
Refactor `StorageMenu` modals to replace deprecated patterns

Saeris committed:
Merge branch 'master' into chore/refactor-storage-modals

Saeris committed:
add test for `DeleteBucketModal` and update test setup

Note: Because this component uses `useParams`, it's necessary to have the dynamic route segment passed to `next-router-mock`'s `createDynamicRouteParser`. In order not to have to manually list all of these as the application grows, I added a glob utility that uses the `pages/` directory  to automatically generate an array of dynamic route paths in this case.

Saeris committed:
add test for `EmptyBucketModal`

Saeris committed:
add test for `EditBucketModal` and add `isNonNullable` utility function

Saeris committed:
add test for `CreateBucketModal`

Saeris committed:
Merge branch 'master' into chore/refactor-storage-modals

Saeris commented on apps/studio/components/interfaces/Storage/__tests__/DeleteBucketModal.test.tsx:
```diff
@@ -0,0 +1,157 @@
+  )
+}
+
+mockAnimationsApi()
```
note that this is needed anytime we've got `DropdownMenu` in the component tree

Saeris replied:
plan to write up some notes there as I discover more edge cases that need special treatment!

Saeris commented on apps/studio/components/interfaces/Storage/__tests__/CreateBucketModal.test.tsx:
```diff
@@ -0,0 +1,99 @@
+
+    fireEvent.click(submitButton)
+
+    await waitFor(() =>
+      expect(routerMock.asPath).toStrictEqual(`/project/default/storage/buckets/test`)
+    )
```
because this component doesn't have an `onClose` callback like the other modals, we instead wait on the route navigation to validate that we've successfully created a new bucket

Saeris commented on apps/studio/components/interfaces/Storage/EditBucketModal.tsx:
```diff
@@ -16,256 +42,332 @@ import { useBucketUpdateMutation } from 'data/storage/bucket-update-mutation'
+    defaultValues: {
+      name: bucket?.name ?? '',
+      public: bucket?.public,
+      has_file_size_limit: isNonNullable(bucket?.file_size_limit),
+      formatted_size_limit: fileSizeLimit ?? 0,
+      allowed_mime_types: (bucket?.allowed_mime_types ?? []).join(', '),
+    },
+    values: {
+      name: bucket?.name ?? '',
+      public: bucket?.public,
+      has_file_size_limit: isNonNullable(bucket?.file_size_limit),
```
at least as far as the OpenAPI response type indicates, `file_size_limit` can be `number | undefined`, so I decided to pull in a handy `isNonNullable` utility here to guard against `null | undefined`. it's useful as a type guard in a variety of contexts (especially for narrowing types) so hopefully it can find broader use elsewhere too

Saeris commented on apps/studio/components/interfaces/Storage/EditBucketModal.tsx:
```diff
@@ -16,256 +42,332 @@ import { useBucketUpdateMutation } from 'data/storage/bucket-update-mutation'
   const formattedGlobalUploadLimit = `${value} ${unit}`
 
-  const [selectedUnit, setSelectedUnit] = useState<StorageSizeUnits>(StorageSizeUnits.BYTES)
+  const [selectedUnit, setSelectedUnit] = useState<string>(StorageSizeUnits.BYTES)
```
note that the type change here is necessary for `Select_shadcn` not to raise a type error

Saeris commented on apps/studio/components/interfaces/Storage/EditBucketModal.tsx:
```diff
@@ -16,256 +42,332 @@ import { useBucketUpdateMutation } from 'data/storage/bucket-update-mutation'
+      open
+      onOpenChange={(open) => {
+        if (!open) {
+          form.reset()
```
thinking about it, `form.reset()` might not be necessary at all because the moment the dialog closes its unmounted from the component tree, which should reset its state implicitly

might come back and update this after testing that theory

Saeris commented on apps/studio/components/interfaces/Storage/EditBucketModal.tsx:
```diff
@@ -16,256 +42,332 @@ import { useBucketUpdateMutation } from 'data/storage/bucket-update-mutation'
+                                <FormControl_Shadcn_>
+                                  <Input_Shadcn_
+                                    id="formatted_size_limit"
+                                    aria-label="File size limit"
```
because we don't wrap this input with a label, we should at least give it an `aria-label` attribute describing what it is for screen readers. it also helps us to target the input with `screen.getByLabelText` in testing

Saeris commented on apps/studio/tests/lib/route-mock.ts:
```diff
@@ -1,11 +1,26 @@
-    '/projects/[ref]',
-  ])
-)
+const pipe =
+  <T>(...fns: Array<(arg: T) => T>) =>
+  (value: T) =>
+    fns.reduce((acc, fn) => fn(acc), value)
+// normalize file paths to posix format (windows FS support)
+const toPosix = (str: string) => str.replaceAll(sep, posix.sep)
+const pagesRoot = toPosix(join(process.cwd(), `./pages`))
+// remove the base filepath
+const removePrefix = (str: string) => str.replace(pagesRoot, ``)
+// remove the file extension or index segment
+const removeExt = (str: string) => str.replace(/.tsx|\/index.tsx/, ``)
+
+// Glob the `pages/` directory for all dynamic route segments
+const paths = [
+  ...(await Array.fromAsync(glob(`${pagesRoot}/**/*\].tsx`))),
+  ...(await Array.fromAsync(glob(`${pagesRoot}/**/**\]/**/index.tsx`))),
+].map(pipe(normalize, toPosix, removePrefix, removeExt))
+
+routerMock.useParser(createDynamicRouteParser(paths))
```
this will make it easier to maintain our router mock by globbing for dynamic routes in our pages directory rather than having to manually maintain that list by hand

the `useParams` hook in particular relies on this list in order to know what route segments are parameters or not

Saeris commented on apps/studio/tests/setup/polyfills.ts:
```diff
@@ -35,3 +35,5 @@ Object.defineProperties(globalThis, {
   TransformStream: { value: TransformStream },
 })
+
+window.HTMLElement.prototype.hasPointerCapture = vi.fn()
```
a click event on `SelectTrigger_shadcn_` calls this browser API, so we need to mock it

Saeris committed:
Merge branch 'master' into chore/refactor-storage-modals

jordienr reviewed (changes requested):
looking good, I think we could add some tests for error scenarios like network errors. 

One UI issue here in the padding and bottom border for the warning. Also similar issue happening in the edit dialog!

<img width="1312" height="1660" alt="Image" src="https://github.com/user-attachments/assets/3025acc4-9424-4c58-a76b-1454d08e4a07" />

prod for reference:

<img width="1500" height="1646" alt="Image" src="https://github.com/user-attachments/assets/1883f50a-2bf6-47a4-be0b-7e8c959cd2cc" />

Saeris commented on apps/studio/components/interfaces/Storage/__tests__/CreateBucketModal.test.tsx:
```diff
@@ -0,0 +1,99 @@
+    addAPIMock({
+      method: `get`,
+      path: `/platform/projects/:ref`,
+      // @ts-expect-error
```
naw, you may have missed that I updated `addApiMock` in another PR to behave like the fetcher client, so this one is legit coming from the openapi response type. the data I'm using is a response I copied from my local, which is missing some required fields specified by the api spec

could make this more permissive in the types there with like `Partial<>`, but otherwise this is the desired behavior because for the most part we want to have errors crop up here as the API spec changes so we know when to expect different behavior to test for

there is also potential to improve this such that we can automatically mock the response data using a tool such as [`zod-mock`](https://github.com/anatine/zod-plugins/tree/main/packages/zod-mock) by leveraging generated response schemas, but that's work for another day

Saeris commented on apps/studio/components/interfaces/Storage/__tests__/CreateBucketModal.test.tsx:
```diff
@@ -0,0 +1,99 @@
+
+describe(`CreateBucketModal`, () => {
+  beforeEach(() => {
+    vi.mock(`hooks/misc/useCheckPermissions`, () => ({
+      useCheckPermissions: vi.fn().mockImplementation(() => true),
+    }))
```
agree that it would be nice to have this in the custom render!

Saeris committed:
Merge branch 'master' into chore/refactor-storage-modals

Saeris committed:
implement requested changes

Saeris committed:
Merge branch 'master' into chore/refactor-storage-modals

Saeris committed:
implement visual fixes

Saeris commented:
> looking good, I think we could add some tests for error scenarios like network errors.
> 
> One UI issue here in the padding and bottom border for the warning. Also similar issue happening in the edit dialog!

Think I fixed this. Issue is that we switched from Modal to Dialog, and Dialog Content has a padding value on it, in addition to Admonition having a padding value of it's own.

Saeris committed:
Merge branch 'master' into chore/refactor-storage-modals
